File: Overlord.Core/Overlord.Core/SpaceCombatSystem.cs
using Overlord.Core.Models;

namespace Overlord.Core;

/// <summary>
/// Platform-agnostic space combat system.
/// Handles fleet-vs-fleet orbital battles, damage calculations, and craft destruction.
/// </summary>
public class SpaceCombatSystem
{
    private readonly GameState _gameState;
    private readonly UpgradeSystem _upgradeSystem;
    private readonly DefenseSystem _defenseSystem;

    /// <summary>
    /// Event fired when space battle starts.
    /// Parameters: (battle)
    /// </summary>
    public event Action<SpaceBattle>? OnSpaceBattleStarted;

    /// <summary>
    /// Event fired when space battle completes.
    /// Parameters: (battle, result)
    /// </summary>
    public event Action<SpaceBattle, SpaceBattleResult>? OnSpaceBattleCompleted;

    /// <summary>
    /// Event fired when craft is destroyed.
    /// Parameters: (craftID, owner)
    /// </summary>
    public event Action<int, FactionType>? OnCraftDestroyed;

    public SpaceCombatSystem(GameState gameState, UpgradeSystem upgradeSystem, DefenseSystem defenseSystem)
    {
        _gameState = gameState ?? throw new ArgumentNullException(nameof(gameState));
        _upgradeSystem = upgradeSystem ?? throw new ArgumentNullException(nameof(upgradeSystem));
        _defenseSystem = defenseSystem ?? throw new ArgumentNullException(nameof(defenseSystem));
    }

    /// <summary>
    /// Initiates a space battle at a planet.
    /// </summary>
    /// <param name="planetID">Planet where battle occurs</param>
    /// <param name="attackerFaction">Attacking faction</param>
    /// <returns>Space battle instance if combat initiated, null if no combat possible</returns>
    public SpaceBattle? InitiateSpaceBattle(int planetID, FactionType attackerFaction)
    {
        // Get planet
        if (!_gameState.PlanetLookup.TryGetValue(planetID, out var planet))
            return null;

        // Get craft at planet
        var allCraft = _gameState.Craft.Where(c => c.PlanetID == planetID && !c.IsDeployed).ToList();

        // Get attacker and defender craft
        var attackerCraft = allCraft.Where(c => c.Owner == attackerFaction).Select(c => c.ID).ToList();
        var defenderCraft = allCraft.Where(c => c.Owner == planet.Owner).Select(c => c.ID).ToList();

        // Need both sides to have craft
        if (attackerCraft.Count == 0 || defenderCraft.Count == 0)
            return null;

        // Check if defender has orbital defenses
        bool hasOrbitalDefense = _defenseSystem.GetActiveDefensePlatformCount(planetID) > 0;

        // Create battle
        var battle = new SpaceBattle
        {
            PlanetID = planetID,
            PlanetName = planet.Name,
            AttackerFaction = attackerFaction,
            DefenderFaction = planet.Owner,
            AttackerCraftIDs = attackerCraft,
            DefenderCraftIDs = defenderCraft,
            DefenderHasOrbitalDefense = hasOrbitalDefense
        };

        OnSpaceBattleStarted?.Invoke(battle);
        return battle;
    }

    /// <summary>
    /// Executes space combat between two fleets.
    /// </summary>
    /// <param name="battle">Battle to execute</param>
    /// <returns>Battle result</returns>
    public SpaceBattleResult ExecuteSpaceCombat(SpaceBattle battle)
    {
        // Get craft entities
        var attackerFleet = battle.AttackerCraftIDs
            .Select(id => _gameState.CraftLookup.TryGetValue(id, out var c) ? c : null)
            .Where(c => c != null)
            .Cast<CraftEntity>()
            .ToList();

        var defenderFleet = battle.DefenderCraftIDs
            .Select(id => _gameState.CraftLookup.TryGetValue(id, out var c) ? c : null)
            .Where(c => c != null)
            .Cast<CraftEntity>()
            .ToList();

        // Calculate fleet strengths
        int attackerStrength = CalculateFleetStrength(attackerFleet, battle.AttackerFaction);
        int defenderStrength = CalculateFleetStrength(defenderFleet, battle.DefenderFaction);

        // Apply orbital defense bonus
        if (battle.DefenderHasOrbitalDefense)
        {
            float defenseMultiplier = _defenseSystem.GetDefenseMultiplier(battle.PlanetID);
            defenderStrength = (int)Math.Floor(defenderStrength * defenseMultiplier);
        }

        // Determine winner (tie goes to attacker)
        bool attackerWins = attackerStrength >= defenderStrength;

        // Calculate damage
        int winnerStrength = attackerWins ? attackerStrength : defenderStrength;
        int loserStrength = attackerWins ? defenderStrength : attackerStrength;
        float strengthRatio = winnerStrength / (float)Math.Max(1, loserStrength);
        int damagePerCraft = (int)Math.Floor((strengthRatio - 1.0f) * 50f);

        // Apply damage to losing fleet
        var losingFleet = attackerWins ? defenderFleet : attackerFleet;
        var craftDamage = new Dictionary<int, int>();
        var destroyedCraftIDs = new List<int>();

        ApplyDamage(losingFleet, damagePerCraft, craftDamage, destroyedCraftIDs);

        // Remove destroyed craft from game state
        foreach (var craftID in destroyedCraftIDs)
        {
            if (_gameState.CraftLookup.TryGetValue(craftID, out var craft))
            {
                _gameState.Craft.Remove(craft);
                OnCraftDestroyed?.Invoke(craftID, craft.Owner);
            }
        }

        _gameState.RebuildLookups();

        // Create result
        var result = new SpaceBattleResult
        {
            AttackerWins = attackerWins,
            AttackerStrength = attackerStrength,
            DefenderStrength = defenderStrength,
            DamagePerCraft = damagePerCraft,
            DestroyedCraftIDs = destroyedCraftIDs,
            CraftDamage = craftDamage
        };

        OnSpaceBattleCompleted?.Invoke(battle, result);
        return result;
    }

    /// <summary>
    /// Calculates total fleet strength.
    /// </summary>
    /// <param name="fleet">List of craft in fleet</param>
    /// <param name="faction">Faction owning the fleet</param>
    /// <returns>Total strength</returns>
    private int CalculateFleetStrength(List<CraftEntity> fleet, FactionType faction)
    {
        int totalStrength = 0;

        // Get weapon damage modifier for faction
        float weaponMod = _upgradeSystem.GetDamageModifier(faction);

        foreach (var craft in fleet)
        {
            int craftStrength = CalculateCraftStrength(craft, weaponMod);
            totalStrength += craftStrength;
        }

        return totalStrength;
    }

    /// <summary>
    /// Calculates strength for a single craft.
    /// </summary>
    /// <param name="craft">Craft to calculate</param>
    /// <param name="weaponMod">Weapon damage modifier (from UpgradeSystem)</param>
    /// <returns>Craft strength</returns>
    private int CalculateCraftStrength(CraftEntity craft, float weaponMod)
    {
        int baseStrength = craft.Type switch
        {
            CraftType.BattleCruiser => 100,
            CraftType.CargoCruiser => 30,
            _ => 0 // Solar Satellite and Atmosphere Processor don't participate in space combat
        };

        // Apply weapon modifier (Battle Cruiser only)
        if (craft.Type == CraftType.BattleCruiser)
        {
            baseStrength = (int)Math.Floor(baseStrength * weaponMod);
        }

        // Apply damage modifier (current HP / max HP)
        float hpPercent = craft.Health / 100f; // Assume max HP is 100
        int strength = (int)Math.Floor(baseStrength * hpPercent);

        return strength;
    }

    /// <summary>
    /// Applies damage to a fleet, distributing evenly across all craft.
    /// Destroys craft when HP ≤ 0.
    /// </summary>
    /// <param name="fleet">Fleet to damage</param>
    /// <param name="damagePerCraft">Damage per craft</param>
    /// <param name="craftDamage">Output: Dictionary of craft ID to damage taken</param>
    /// <param name="destroyedCraftIDs">Output: List of destroyed craft IDs</param>
    private void ApplyDamage(List<CraftEntity> fleet, int damagePerCraft, Dictionary<int, int> craftDamage, List<int> destroyedCraftIDs)
    {
        foreach (var craft in fleet)
        {
            // Apply damage
            craft.Health -= damagePerCraft;
            craftDamage[craft.ID] = damagePerCraft;

            // Check if destroyed
            if (craft.Health <= 0)
            {
                destroyedCraftIDs.Add(craft.ID);
            }
        }
    }

    /// <summary>
    /// Gets all hostile craft at a planet (for detecting space battles).
    /// </summary>
    /// <param name="planetID">Planet ID</param>
    /// <param name="faction">Faction to check against</param>
    /// <returns>List of hostile craft IDs</returns>
    public List<int> GetHostileCraftAtPlanet(int planetID, FactionType faction)
    {
        if (!_gameState.PlanetLookup.TryGetValue(planetID, out var planet))
            return new List<int>();

        return _gameState.Craft
            .Where(c => c.PlanetID == planetID && c.Owner != faction && !c.IsDeployed)
            .Select(c => c.ID)
            .ToList();
    }

    /// <summary>
    /// Checks if a space battle should occur at a planet.
    /// </summary>
    /// <param name="planetID">Planet ID</param>
    /// <returns>True if hostile fleets are present</returns>
    public bool ShouldSpaceBattleOccur(int planetID)
    {
        if (!_gameState.PlanetLookup.TryGetValue(planetID, out var planet))
            return false;

        var craft = _gameState.Craft.Where(c => c.PlanetID == planetID && !c.IsDeployed).ToList();

        // Check if multiple factions have craft at this planet
        var factions = craft.Select(c => c.Owner).Distinct().ToList();
        return factions.Count > 1; // Battle if multiple factions present
    }
}

File: Overlord.Phaser/src/core/SpaceCombatSystem.ts
import { GameState } from './GameState';
import { UpgradeSystem } from './UpgradeSystem';
import { DefenseSystem } from './DefenseSystem';
import { FactionType, CraftType } from './models/Enums';
import { SpaceBattle, SpaceBattleResult } from './models/CombatModels';
import { CraftEntity } from './models/CraftEntity';

/**
 * Platform-agnostic space combat system.
 * Handles fleet-vs-fleet orbital battles, damage calculations, and craft destruction.
 */
export class SpaceCombatSystem {
  private readonly gameState: GameState;
  private readonly upgradeSystem: UpgradeSystem;
  private readonly defenseSystem: DefenseSystem;

  /**
   * Event fired when space battle starts.
   * Parameters: (battle)
   */
  public onSpaceBattleStarted?: (battle: SpaceBattle) => void;

  /**
   * Event fired when space battle completes.
   * Parameters: (battle, result)
   */
  public onSpaceBattleCompleted?: (battle: SpaceBattle, result: SpaceBattleResult) => void;

  /**
   * Event fired when craft is destroyed.
   * Parameters: (craftID, owner)
   */
  public onCraftDestroyed?: (craftID: number, owner: FactionType) => void;

  constructor(gameState: GameState, upgradeSystem: UpgradeSystem, defenseSystem: DefenseSystem) {
    if (!gameState) {
      throw new Error('gameState cannot be null or undefined');
    }
    if (!upgradeSystem) {
      throw new Error('upgradeSystem cannot be null or undefined');
    }
    if (!defenseSystem) {
      throw new Error('defenseSystem cannot be null or undefined');
    }

    this.gameState = gameState;
    this.upgradeSystem = upgradeSystem;
    this.defenseSystem = defenseSystem;
  }

  /**
   * Initiates a space battle at a planet.
   * @param planetID Planet where battle occurs
   * @param attackerFaction Attacking faction
   * @returns Space battle instance if combat initiated, null if no combat possible
   */
  public initiateSpaceBattle(planetID: number, attackerFaction: FactionType): SpaceBattle | null {
    // Get planet
    const planet = this.gameState.planetLookup.get(planetID);
    if (!planet) {
      return null;
    }

    // Get craft at planet (excluding deployed craft like Solar Satellites)
    const allCraft = this.gameState.craft.filter(c => c.planetID === planetID && !c.isDeployed);

    // Get attacker and defender craft
    const attackerCraft = allCraft.filter(c => c.owner === attackerFaction).map(c => c.id);
    const defenderCraft = allCraft.filter(c => c.owner === planet.owner).map(c => c.id);

    // Need both sides to have craft
    if (attackerCraft.length === 0 || defenderCraft.length === 0) {
      return null;
    }

    // Check if defender has orbital defenses
    const hasOrbitalDefense = this.defenseSystem.getActiveDefensePlatformCount(planetID) > 0;

    // Create battle
    const battle = new SpaceBattle();
    battle.planetID = planetID;
    battle.planetName = planet.name;
    battle.attackerFaction = attackerFaction;
    battle.defenderFaction = planet.owner;
    battle.attackerCraftIDs = attackerCraft;
    battle.defenderCraftIDs = defenderCraft;
    battle.defenderHasOrbitalDefense = hasOrbitalDefense;

    this.onSpaceBattleStarted?.(battle);
    return battle;
  }

  /**
   * Executes space combat between two fleets.
   * @param battle Battle to execute
   * @returns Battle result
   */
  public executeSpaceCombat(battle: SpaceBattle): SpaceBattleResult {
    // Get craft entities
    const attackerFleet = battle.attackerCraftIDs
      .map(id => this.gameState.craftLookup.get(id))
      .filter(c => c !== undefined) as CraftEntity[];

    const defenderFleet = battle.defenderCraftIDs
      .map(id => this.gameState.craftLookup.get(id))
      .filter(c => c !== undefined) as CraftEntity[];

    // Calculate fleet strengths
    let attackerStrength = this.calculateFleetStrength(attackerFleet, battle.attackerFaction);
    let defenderStrength = this.calculateFleetStrength(defenderFleet, battle.defenderFaction);

    // Apply orbital defense bonus
    if (battle.defenderHasOrbitalDefense) {
      const defenseMultiplier = this.defenseSystem.getDefenseMultiplier(battle.planetID);
      defenderStrength = Math.floor(defenderStrength * defenseMultiplier);
    }

    // Determine winner (tie goes to attacker)
    const attackerWins = attackerStrength >= defenderStrength;

    // Calculate damage
    const winnerStrength = attackerWins ? attackerStrength : defenderStrength;
    const loserStrength = attackerWins ? defenderStrength : attackerStrength;
    const strengthRatio = winnerStrength / Math.max(1, loserStrength);
    const damagePerCraft = Math.floor((strengthRatio - 1.0) * 50);

    // Apply damage to losing fleet
    const losingFleet = attackerWins ? defenderFleet : attackerFleet;
    const craftDamage = new Map<number, number>();
    const destroyedCraftIDs: number[] = [];

    this.applyDamage(losingFleet, damagePerCraft, craftDamage, destroyedCraftIDs);

    // Remove destroyed craft from game state
    for (const craftID of destroyedCraftIDs) {
      const craft = this.gameState.craftLookup.get(craftID);
      if (craft) {
        const index = this.gameState.craft.indexOf(craft);
        if (index !== -1) {
          this.gameState.craft.splice(index, 1);
        }
        this.onCraftDestroyed?.(craftID, craft.owner);
      }
    }

    this.gameState.rebuildLookups();

    // Create result
    const result = new SpaceBattleResult();
    result.attackerWins = attackerWins;
    result.attackerStrength = attackerStrength;
    result.defenderStrength = defenderStrength;
    result.damagePerCraft = damagePerCraft;
    result.destroyedCraftIDs = destroyedCraftIDs;
    result.craftDamage = craftDamage;

    this.onSpaceBattleCompleted?.(battle, result);
    return result;
  }

  /**
   * Calculates total fleet strength.
   * @param fleet List of craft in fleet
   * @param faction Faction owning the fleet
   * @returns Total strength
   */
  private calculateFleetStrength(fleet: CraftEntity[], faction: FactionType): number {
    let totalStrength = 0;

    // Get weapon damage modifier for faction
    const weaponMod = this.upgradeSystem.getDamageModifier(faction);

    for (const craft of fleet) {
      const craftStrength = this.calculateCraftStrength(craft, weaponMod);
      totalStrength += craftStrength;
    }

    return totalStrength;
  }

  /**
   * Calculates strength for a single craft.
   * @param craft Craft to calculate
   * @param weaponMod Weapon damage modifier (from UpgradeSystem)
   * @returns Craft strength
   */
  private calculateCraftStrength(craft: CraftEntity, weaponMod: number): number {
    let baseStrength: number;

    switch (craft.type) {
      case CraftType.BattleCruiser:
        baseStrength = 100;
        break;
      case CraftType.CargoCruiser:
        baseStrength = 30;
        break;
      default:
        baseStrength = 0; // Solar Satellite and Atmosphere Processor don't participate in space combat
    }

    // Apply weapon modifier (Battle Cruiser only)
    if (craft.type === CraftType.BattleCruiser) {
      baseStrength = Math.floor(baseStrength * weaponMod);
    }

    // Apply damage modifier (current HP / max HP)
    const hpPercent = craft.health / 100; // Assume max HP is 100
    const strength = Math.floor(baseStrength * hpPercent);

    return strength;
  }

  /**
   * Applies damage to a fleet, distributing evenly across all craft.
   * Destroys craft when HP ≤ 0.
   * @param fleet Fleet to damage
   * @param damagePerCraft Damage per craft
   * @param craftDamage Output: Map of craft ID to damage taken
   * @param destroyedCraftIDs Output: List of destroyed craft IDs
   */
  private applyDamage(
    fleet: CraftEntity[],
    damagePerCraft: number,
    craftDamage: Map<number, number>,
    destroyedCraftIDs: number[]
  ): void {
    for (const craft of fleet) {
      // Apply damage
      craft.health -= damagePerCraft;
      craftDamage.set(craft.id, damagePerCraft);

      // Check if destroyed
      if (craft.health <= 0) {
        destroyedCraftIDs.push(craft.id);
      }
    }
  }

  /**
   * Gets all hostile craft at a planet (for detecting space battles).
   * @param planetID Planet ID
   * @param faction Faction to check against
   * @returns List of hostile craft IDs
   */
  public getHostileCraftAtPlanet(planetID: number, faction: FactionType): number[] {
    const planet = this.gameState.planetLookup.get(planetID);
    if (!planet) {
      return [];
    }

    return this.gameState.craft
      .filter(c => c.planetID === planetID && c.owner !== faction && !c.isDeployed)
      .map(c => c.id);
  }

  /**
   * Checks if a space battle should occur at a planet.
   * @param planetID Planet ID
   * @returns True if hostile fleets are present
   */
  public shouldSpaceBattleOccur(planetID: number): boolean {
    const planet = this.gameState.planetLookup.get(planetID);
    if (!planet) {
      return false;
    }

    const craft = this.gameState.craft.filter(c => c.planetID === planetID && !c.isDeployed);

    // Check if multiple factions have craft at this planet
    const factions = new Set(craft.map(c => c.owner));
    return factions.size > 1; // Battle if multiple factions present
  }
}
